python/ray/actor.py

```diff
@@ -93,73 +100,123 @@ def get_actor_checkpoint(worker, actor_id):
+    return worker.redis_client.hmget(
+        actor_key, ["checkpoint_index", "checkpoint", "frontier"])
 
 
-def make_actor_method_executor(worker, method_name, method):
+def save_and_log_checkpoint(worker, actor):
+    """Save a checkpoint on the actor and log any errors.
+
+    Args:
+        worker: The worker to use to log errors.
+        actor: The actor to checkpoint.
+        checkpoint_index: The number of tasks that have executed so far.
+    """
+    try:
+        actor.__ray_checkpoint__()
+    except:
+        traceback_str = ray.utils.format_error_message(
+            traceback.format_exc())
+        # Log the error message.
+        ray.utils.push_error_to_driver(
+            worker.redis_client,
+            "checkpoint",
+            traceback_str,
+            driver_id=worker.task_driver_id.id(),
+            data={"actor_class": actor.__class__.__name__,
+                  "function_name": actor.__ray_checkpoint__.__name__})
+
+
+def restore_and_log_checkpoint(worker, actor):
+    """Restore an actor from a checkpoint and log any errors.
+
+    Args:
+        worker: The worker to use to log errors.
+        actor: The actor to restore.
+    """
+    checkpoint_resumed = False
+    try:
+        checkpoint_resumed = actor.__ray_checkpoint_restore__()
+    except:
+        traceback_str = ray.utils.format_error_message(
+            traceback.format_exc())
+        # Log the error message.
+        ray.utils.push_error_to_driver(
+            worker.redis_client,
+            "checkpoint",
+            traceback_str,
+            driver_id=worker.task_driver_id.id(),
+            data={
+                "actor_class": actor.__class__.__name__,
+                "function_name":
+                actor.__ray_checkpoint_restore__.__name__})
+    return checkpoint_resumed
+
+
+def make_actor_method_executor(worker, method_name, method, actor_imported):
     """Make an executor that wraps a user-defined actor method.
 
-    The executor wraps the method to update the worker's internal state. If the
-    task is a success, the dummy object returned is added to the object store,
-    to signal that the following task can run, and the worker's task counter is
-    updated to match the executed task. Else, the executor reports failure to
-    the local scheduler so that the task counter does not get updated.
+    The wrapped method updates the worker's internal state and performs any
+    necessary checkpointing operations.
 
     Args:
         worker (Worker): The worker that is executing the actor.
         method_name (str): The name of the actor method.
         method (instancemethod): The actor method to wrap. This should be a
             method defined on the actor class and should therefore take an
             instance of the actor as the first argument.
+        actor_imported (bool): Whether the actor has been imported.
+            Checkpointing operations will not be run if this is set to False.
```

Comment:
I see, this bool is just to avoid loading checkpoints in the case where unpickling the actor class raises an exception, right?

Reply:
Yes, there was an error where the checkpointing methods couldn't be found if the actor hasn't been imported yet.